### ISSUE TITLE
How to run multiple consumers?

### ISSUE THREAD

**Issue by Paxa** [zendesk/racecar#21]
It support only 1 consumer right now, is it?
Any plans to add support for multiple consumers?

---

I have few more questions:

What is usual practice to make retries? I don't find any functionality to access publishing messages, or access to kafka client object (it'a a local variable)

Exception in `teardown` (defined in consumer) will make process exit immediately, will it be better to call `Racecar.config.error_handler` ?

Is it possible not to load rails on load? If we want to run plain ruby consumer inside rails project

**Comment by Paxa** [comment]
I just got exception that made all process stop:
```
I, [2017-08-19T01:29:23.967512 #46084]  INFO -- : Committing offsets: mrt-retry/5:37, mrt-retry/14:27, mrt-retry/2:29, mrt-retry/17:41, mrt-retry/16:37, mrt-retry/8:36, mrt-retry/0:44, mrt-retry/10:37
D
D, [2017-08-19T01:29:23.967626 #46084] DEBUG -- : Sending request 56 to kafka2.stg.veritrans.co.id:9092
D, [2017-08-19T01:29:23.968113 #46084] DEBUG -- : Waiting for response 56 from kafka2.stg.veritrans.co.id:9092
D, [2017-08-19T01:29:24.046070 #46084] DEBUG -- : Received response 56 from kafka2.stg.veritrans.co.id:9092
E, [2017-08-19T01:29:24.046420 #46084] ERROR -- : Error committing offsets: Kafka::IllegalGeneration
E, [2017-08-19T01:29:24.046787 #46084] ERROR -- : Consumer thread crashed: Kafka::OffsetCommitError: Kafka::IllegalGeneration
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/bundler/gems/ruby-kafka-7f47bbe8c716/lib/kafka/consumer_group.rb:83:in `rescue in commit_offsets'
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/bundler/gems/ruby-kafka-7f47bbe8c716/lib/kafka/consumer_group.rb:68:in `commit_offsets'
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/bundler/gems/ruby-kafka-7f47bbe8c716/lib/kafka/offset_manager.rb:67:in `commit_offsets'
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/bundler/gems/ruby-kafka-7f47bbe8c716/lib/kafka/consumer.rb:342:in `make_final_offsets_commit!'
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/bundler/gems/ruby-kafka-7f47bbe8c716/lib/kafka/consumer.rb:336:in `ensure in consumer_loop'
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/bundler/gems/ruby-kafka-7f47bbe8c716/lib/kafka/consumer.rb:338:in `consumer_loop'
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/bundler/gems/ruby-kafka-7f47bbe8c716/lib/kafka/consumer.rb:182:in `each_message'
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/bundler/gems/racecar-466a27262b36/lib/racecar/runner.rb:53:in `run'
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/bundler/gems/racecar-466a27262b36/lib/racecar.rb:27:in `run'
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/bundler/gems/racecar-466a27262b36/lib/racecar/cli.rb:52:in `main'
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/bundler/gems/racecar-466a27262b36/exe/racecar:8:in `<top (required)>'
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/bin/racecar:22:in `load'
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/bin/racecar:22:in `<top (required)>'
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/bundler-1.15.3/lib/bundler/cli/exec.rb:74:in `load'
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/bundler-1.15.3/lib/bundler/cli/exec.rb:74:in `kernel_load'
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/bundler-1.15.3/lib/bundler/cli/exec.rb:27:in `run'
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/bundler-1.15.3/lib/bundler/cli.rb:365:in `exec'
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/bundler-1.15.3/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/bundler-1.15.3/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/bundler-1.15.3/lib/bundler/vendor/thor/lib/thor.rb:369:in `dispatch'
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/bundler-1.15.3/lib/bundler/cli.rb:22:in `dispatch'
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/bundler-1.15.3/lib/bundler/vendor/thor/lib/thor/base.rb:444:in `start'
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/bundler-1.15.3/lib/bundler/cli.rb:13:in `start'
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/bundler-1.15.3/exe/bundle:30:in `block in <top (required)>'
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/bundler-1.15.3/lib/bundler/friendly_errors.rb:121:in `with_friendly_errors'
/Users/pavel/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/bundler-1.15.3/exe/bundle:22:in `<top (required)>'
/Users/pavel/.rbenv/versions/2.4.0/bin/bundle:22:in `load'
/Users/pavel/.rbenv/versions/2.4.0/bin/bundle:22:in `<main>'
I, [2017-08-19T01:29:24.047323 #46084]  INFO -- : Gracefully shutting down
=> Shut down
```

Is it expected?

**Comment by filiptepper** [comment]
> It support only 1 consumer right now, is it?
> Any plans to add support for multiple consumers?

What do you mean by 1 consumer? `racecar` supports partition balancing and it works as expected. Here I'm running 3 instances of the same consumer in separate Docker instances, each consuming a different partition.

<img width="924" alt="screen shot 2017-08-21 at 11 15 38" src="https://user-images.githubusercontent.com/3822/29512330-2daeb6f0-8662-11e7-8dd0-16e7eab64e29.png">

> What is usual practice to make retries? I don't find any functionality to access publishing messages, or access to kafka client object (it'a a local variable)

That depends on your use case and preferences. Most of my consumers work in batches, if batch fails I always want to retry. I haven't really ported any other types of consumers to `racecar` yet so I can't be much of an assistance here.

> Exception in teardown (defined in consumer) will make process exit immediately, will it be better to call Racecar.config.error_handler ?

This sounds interesting, what do you think @dasch?

> Is it possible not to load rails on load? If we want to run plain ruby consumer inside rails project

Yes, I use `racecar` without Rails.

```bash
$ bundle exec racecar --help
Usage: racecar MyConsumer [options]
    -r, --require LIBRARY            Require the LIBRARY before starting the consumer
        --version                    Show Racecar version
```

All you need is a file that you `--require` which defines `MyConsumer` class (or any other to your liking).

**Comment by dasch** [comment]
@filiptepper wrote:
> > What is usual practice to make retries? I don't find any functionality to access publishing messages, or access to kafka client object (it'a a local variable)
> 
> That depends on your use case and preferences. Most of my consumers work in batches, if batch fails I always want to retry. I haven't really ported any other types of consumers to `racecar` yet so I can't be much of an assistance here.

In our use, exceptions in the processing code would either indicate a failure to correctly handle some specific message or the temporary unavailability of some external network resource. In each case, we want to retry _slowly_, i.e. without using up all the CPU resources on the host and without overwhelming the external resource. The way we do that is by _pausing_ the partition from which the message came for some number of seconds, which is the same as backing off before retrying. If the `pause_timeout` config variable is set to 10 seconds, for instance, you'll do up to 6 retries a minute.

> > Exception in teardown (defined in consumer) will make process exit immediately, will it be better to call Racecar.config.error_handler ?
> 
> This sounds interesting, what do you think @dasch?

What would the use case be? What kind of error handlers are you using?

> > Is it possible not to load rails on load? If we want to run plain ruby consumer inside rails project
> 
> Yes, I use `racecar` without Rails.
> 
> ```bash
> $ bundle exec racecar --help
> Usage: racecar MyConsumer [options]
>     -r, --require LIBRARY            Require the LIBRARY before starting the consumer
>         --version                    Show Racecar version
> ```
> 
> All you need is a file that you `--require` which defines `MyConsumer` class (or any other to your liking).

Racecar currently assumes that you want to load Rails when in a Rails project – I think it's a sane assumption, and shouldn't really be a problem – you can just elect to not use the Rails bits in your consumer.

**Comment by dasch** [comment]
@Paxa wrote:
> It support only 1 consumer right now, is it?
> Any plans to add support for multiple consumers?

Racecar supports as many consumers as you like, you just have to start the `racecar` process with the same consumer class `n` times in order to get `n` consumers in your group. In our typical setup, we run 6 instances of `racecar` spread over 3 hosts.

**Comment by Paxa** [comment]
By multiple consumers I mean processing different messages from different topics with different ruby classes.
Something like:
```
bundle exec racecar MyConsumer1,MyConsumer2
```

----

> Racecar currently assumes that you want to load Rails when in a Rails project – I think it's a sane assumption

It is, don't need it much, just nice to have

----

About retries: let's say one message cause an exception and we reject a message, will it process newer messages in that partition or keep retrying that one?

**Comment by dasch** [comment]
@Paxa wrote:
> By multiple consumers I mean processing different messages from different topics with different ruby classes.
> Something like:
> ```
> bundle exec racecar MyConsumer1,MyConsumer2
> ```

That's not supported, and not really necessary: just execute `racecar` multiple times, once for each consumer class:

    $ bundle exec racecar MyConsumer1

And then in another terminal (or preferably, with another Runit service or the like):

    $ bundle exec racecar MyConsumer2

Each consumer will run in its own process, providing much better isolation.

> About retries: let's say one message cause an exception and we reject a message, will it process newer messages in that partition or keep retrying that one?

I added a section to the README concerning this – does that adequately answer your question?
